crates/vm/src/arch/interpreter.rs

```diff
@@ -405,22 +409,38 @@ where
     ) -> Result<(Vec<Segment>, VmState<F, GuestMemory>), ExecutionError> {
         let mut exec_state = VmExecState::new(from_state, ctx);
 
-        let instret = exec_state.instret();
-        let pc = exec_state.pc();
-        let segmentation_check_insns = exec_state.ctx.segmentation_ctx.segment_check_insns;
+        loop {
+            exec_state = self.execute_metered_from_state_until_suspension(exec_state)?;
+            // The execution has terminated.
+            if exec_state.exit_code.is_ok() && exec_state.exit_code.as_ref().unwrap().is_some() {
+                break;
+            }
+            if exec_state.exit_code.is_err() {
+                return Err(exec_state.exit_code.unwrap_err());
+            }
+        }
+        check_termination(exec_state.exit_code)?;
+        let VmExecState { vm_state, ctx, .. } = exec_state;
+        Ok((ctx.into_segments(), vm_state))
+    }
+    pub fn execute_metered_from_state_until_suspension(
+        &self,
+        mut from_state: VmExecState<F, GuestMemory, MeteredCtx>,
```

Comment:
small nit - `from_state` doesn't seem the right term since it's mutated in place